treesit-auto.el

```diff
@@ -112,11 +121,71 @@ remap the tree-sitter variant back to the default mode."
            (when name-mode-bound-p
              (add-to-list 'major-mode-remap-alist `(,name-ts-mode . ,name-mode)))))))
 
+(defun treesit-auto--prompt-to-install-package (lang &optional ignore-prompt)
+  "Ask the user if they want to install a treesitter grammar for `LANG'.
+
+If `IGNORE-PROMPT' is set to true, no prompt will be given and we
+will try to install the tree-sitter grammar as long as a
+repository can be found in `treesit-language-source-alist'.
+
+Returns `non-nil' if install was completed without error."
+  (if-let* ((repo (alist-get lang treesit-language-source-alist))
+            (response (or ignore-prompt
+                          (yes-or-no-p (format "Tree Sitter grammar for %s is missing.  Would you like to install it from: %s?"
+                                               (symbol-name lang)
+                                               (car repo))))))
+      ;; treesit-install-language-grammar will return nil if the
+      ;; operation succeeded and 't if a warning was tossed. I don't
+      ;; think this is by design but just because of the way
+      ;; `display-warning' works.
+      (progn
+        (message "Installing the tree-sitter grammar for %s" lang)
+        (not (treesit-install-language-grammar lang)))))
+
+(defun treesit-auto--maybe-install-grammar ()
+  "Try to install the grammar matching the current major-mode.
+
+If the tree-sitter grammar is missing for the current major mode,
+it will prompt the user if they want to install it from the
+currently registered repository.  If the user chooses to install
+the grammar it will then re-enable the current major-mode."
+  (when-let* ((mode (symbol-name major-mode))
+              (lang (and (string-match "\\(.*\\)-ts-mode$" mode)
+                         (intern (replace-regexp-in-string
+                                  "\\(.*\\)-ts-mode$" "\\1"
+                                  mode))))
+              ((not (treesit-ready-p lang 't)))
```

Comment:
Question: This will only activate if the Tree-sitter mode is activated, right?  Assuming we've made a fallback to the original mode, then `major-mode` would read `python-mode`, instead of `python-ts-mode`.

Future enhancement: I like the code as-is to handle the case of booting up a `-ts-mode` and just having it handle the grammar installation for our first step.  However, I'd eventually like to support an even better UX:

When I land in a fallback mode, I would like an option for it to notify me that a tree-sitter mode is available, if I want to download it.  My configuration, assuming these options exist, would look like this:

```elisp
(setq treesit-auto-install 'prompt)
(global-treesit-auto-mode)
```

Then, when I open a python file and it falls back to `python-mode`, I'd like Emacs to ask about installing the Tree-sitter grammar for Python, to which I would reply "yes" if I have an internet connection.  It's hard to tell if we want to try reworking that now, or do this simple base behavior first, and add the base-mode behavior in another PR afterwards.  Your thoughts?

Reply:
yeah, auto upgrading would be a really nice feature